gossipd/seeker.c

```diff
@@ -0,0 +1,991 @@
+	 * missing channels. */
+	bool unknown_nodes;
+
+	/* Peers we've asked to stream us gossip */
+	struct peer *gossiper_softref[3];
+
+	/* A peer that told us about unknown gossip. */
+	struct peer *preferred_peer_softref;
+
+};
+
+/* Mutual recursion */
+static void seeker_check(struct seeker *seeker);
+static void probe_some_random_scids(struct seeker *seeker);
+
+static void begin_check_timer(struct seeker *seeker)
+{
+	const u32 polltime = GOSSIP_SEEKER_INTERVAL(seeker);
+
+	seeker->check_timer = new_reltimer(&seeker->daemon->timers,
+					   seeker,
+					   time_from_sec(polltime),
+					   seeker_check, seeker);
+}
+
+#if DEVELOPER
+static void memleak_help_seeker(struct htable *memtable,
+				struct seeker *seeker)
+{
+	memleak_remove_htable(memtable, &seeker->unknown_scids.raw);
+	memleak_remove_htable(memtable, &seeker->stale_scids.raw);
+}
+#endif /* DEVELOPER */
+
+#define set_state(seeker, state) \
+	set_state_((seeker), (state), stringify(state))
+
+static void set_state_(struct seeker *seeker, enum seeker_state state,
+		       const char *statename)
+{
+	status_debug("seeker: state = %s", statename);
+	seeker->state = state;
+}
+
+struct seeker *new_seeker(struct daemon *daemon)
+{
+	struct seeker *seeker = tal(daemon, struct seeker);
+
+	seeker->daemon = daemon;
+	scid_map_init(&seeker->unknown_scids);
+	stale_scid_map_init(&seeker->stale_scids);
+	seeker->random_peer_softref = NULL;
+	for (size_t i = 0; i < ARRAY_SIZE(seeker->gossiper_softref); i++)
+		seeker->gossiper_softref[i] = NULL;
+	seeker->preferred_peer_softref = NULL;
+	seeker->unknown_nodes = false;
+	set_state(seeker, STARTING_UP);
+	begin_check_timer(seeker);
+	memleak_add_helper(seeker, memleak_help_seeker);
+	return seeker;
+}
+
+/* Set this peer as our random peer; return false if NULL. */
+static bool selected_peer(struct seeker *seeker, struct peer *peer)
+{
+	if (!peer)
+		return false;
+
+	set_softref(seeker, &seeker->random_peer_softref, peer);
+	status_debug("seeker: chose peer %s",
+		     type_to_string(tmpctx, struct node_id, &peer->id));
+
+	/* Give it some grace in case we immediately hit timer */
+	seeker->prev_gossip_count
+		= peer->gossip_counter - GOSSIP_SEEKER_INTERVAL(seeker);
+	return true;
+}
+
+static void set_preferred_peer(struct seeker *seeker, struct peer *peer)
+{
+	if (seeker->preferred_peer_softref
+	    && seeker->preferred_peer_softref != peer) {
+		clear_softref(seeker, &seeker->preferred_peer_softref);
+		set_softref(seeker, &seeker->preferred_peer_softref, peer);
+	}
+}
+
+/* Get a random peer, but try our preferred peer first, if any.  This
+ * biasses us to the peer that told us of unexpected gossip. */
+static struct peer *random_seeker(struct seeker *seeker,
+				  bool (*check_peer)(const struct peer *peer))
+{
+	struct peer *peer = seeker->preferred_peer_softref;
+
+	/* 80% chance of just choosing a peer who reported the missing stuff */
+	if (peer && check_peer(peer) && pseudorand(5) != 0) {
```

Comment:
where do these 80% odds come from? not sure i follow the connection with the preferred peer selection.

if there's 3 peers, wouldn't the odds be closer to 66%, or are we weighting more heavily towards it since there's a good chance it was inn operation at the time?

Reply:
Added comment:

```C

	/* 80% chance of immediately choosing a peer who reported the missing
	 * stuff: they presumably can tell us more about it.  We don't
	 * *always* choose it because it could be simply spamming us with
	 * invalid announcements to get chosen, and we don't handle that case
	 * well yet. */
```